python/ray/train/lint/check_circular_imports.py

```diff
@@ -0,0 +1,423 @@
+
+    Args:
+        base_train_patching_imports: A dictionary mapping base train module names to their imports.
+        patch_dir: The directory path containing patch train modules.
+
+    Returns:
+        A list of strings describing any circular import violations found.
+    """
+    violations: List[str] = []
+
+    # Get the imports from the patch train init files
+    patch_train_init_files = list(patch_dir.rglob("__init__.py"))
+    patch_train_init_imports = get_file_module_imports(
+        patch_train_init_files, module_match_string="ray.train"
+    )
+
+    # Expand the imports to include reexports
+    expand_to_include_reexports(base_train_patching_imports)
+
+    # Process each patch train init module
+    for base_train_init_module, imports in base_train_patching_imports.items():
+
+        # Get the imports from the patch train files
+        patch_train_files = convert_to_file_paths(imports)
+        patch_train_file_imports = get_file_module_imports(
+            patch_train_files, module_match_string="ray.train"
+        )
+
+        for patch_module, imports in patch_train_file_imports.items():
+            # Skip if the base train init module is in the import path of the patch module
+            if does_overlap(patch_module, base_train_init_module):
+                continue
+
+            # Skip if the patch train module init file imports the base train init module
+            patch_init_module = (
+                ".".join(patch_module.split(".")[:-1])
+                if not is_package(patch_module)
+                else patch_module
+            )
+            patch_init_imports = patch_train_init_imports.get(patch_init_module, [])
+            if any(
+                does_overlap(imp.module, base_train_init_module)
+                for imp in patch_init_imports
+            ):
+                continue
+
+            for patch_import in imports:
+                # If any of those v1 imports go through the init file, then it is a violation
+                if does_overlap(patch_import.module, base_train_init_module):
+                    violations.append(
+                        f"circular-import-train: Circular import between {base_train_init_module} (importing {patch_module}) and {patch_module} (importing {patch_import.module}). Resolve by importing {base_train_init_module} in the __init__.py of {patch_init_module}."
+                    )
+
+    return violations
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--patch_dir",
+        default="ray/train/v2",
+        help="Path to the directory containing patching contents",
+    )
+    args = parser.parse_args()
+
+    # Compute relevant paths
+    base_dir = get_base_dir()
+    base_train_dir = get_base_train_dir()
+    patch_train_dir = base_dir / Path(args.patch_dir)
+
+    # Initialize train packages
+    initialize_train_packages(base_train_dir, patch_train_dir)
```

Comment:
Add more comments to explain what the purpose of each step here. E.g. not obvious why we need to initialize all the packages.

General criteria for this should be that I can read the docs at the top of the file and the `main` method and get a sense of correctness for this file, without having to read the implementation of each individual method.